### PR TITLE
[Unity][Transform] Implement UpdateParamStructInfo

### PR DESCRIPTION
**Pull request by Lunderberg**
Provide a convenience method to update parameter struct info, propagating any changes to internal bindings and return type.

### CODE REVIEW

**Review by slyubomirsky — CHANGES_REQUESTED** [review]
This seems reasonable, but I found it a little difficult to be certain of the logic. You are using `VisitVarDef` but are ignoring vars defined inside `SeqExpr`s, so I take it that this will update only function parameters (per the name). Is that correct? It should perhaps be pointed out. I'm not sure if it would be simpler to override the function node case instead of `VisitVarDef`.

File: src/relax/transform/update_param_struct_info.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  Expr VisitExpr_(const SeqExprNode* op) override {
+    bool cache = inside_expr_;
+    inside_expr_ = true;
+    auto ret = ExprMutator::VisitExpr_(op);
+    inside_expr_ = cache;
+    return ret;
+  }
+
+  TypedPackedFunc<Optional<StructInfo>(Var)> sinfo_func_;
+  bool inside_expr_{false};
```
[the remark this answers is not in the document]
Comment (Lunderberg) [reply src/relax/transform/update_param_struct_info.cc:67]: Good point, it probably would have been better to name it as `inside_function_body_`.  It has been removed in the updated implementation.

**Comment by Lunderberg** [comment]
> You are using VisitVarDef but are ignoring vars defined inside SeqExprs, so I take it that this will update only function parameters (per the name). Is that correct?

That is correct.  Only parameters would be updated, and variable bindings within the function body would have their struct info inferred.

> I'm not sure if it would be simpler to override the function node case instead of VisitVarDef.

I initially ran into some issues with overriding the `Expr VisitExpr_(const FunctionNode*)` mutator, mostly because it was difficult without copy-pasting the entire `ExprMutator` implementation.  In order to delegate to the base class's mutator, it would need to first construct a `Function` object with updated parameters, and then pass that to the base class.  However, that ran into issues with the validity-checking in the `Function` constructor, since the body of the function wasn't yet updated.

Taking another look at it, it looks like it can be moved to the `VisitExpr_(const FunctionNode*)` by using `func.CopyOnWrite()` to avoid going through the normal constructor.  I agree that this is simpler, and have updated the PR to use the less roundabout implementation.

**Review by slyubomirsky — APPROVED** [review]
Thank you for the changes.

**Pull request merged**